text/2019-11-13-follower-replication.md

```diff
@@ -0,0 +1,127 @@
+
+### The Delegate and Commission
+
+As the `Groups` is configured, the leader is able to choose a group member as a *delegate* to send entries to one or the rest group members in Log Replication, which is called Follower Replication. To implement Follower Replication, we basically need three main steps:
+
+1. The leader picks a group member as a delegate of the group
+2. The leader prepares and sends some *commission*s that indicate what entries the picked delegate will send to others by a new message type MsgBroadcast.
+3. The delegate receives this message and executes the commissions
+
+Here is a diagram showing how Follower Replication works:
+
+```
+                                        +-----------+
+                                        |           |
+       +--------------------------------+     C     |
+       |           MsgAppendResp        |           |
+       |                                +-----^-----+
+       |                                      |
+       |                                      |  MsgAppend(e[4,5])
+       |                                      |
++------v-----+                          +-----+-----+
+|            |                          |           |
+|   leader   +-------------------------->     B     | delegate
+|            |   MsgBroadcast{          |           |
++------^-----+     e[2,5]               +-----+-----+
+       |           Commission(C, [4,5])       |
+       |           Commission(D, [3,5])       |  MsgAppend(e[3,5])
+       |         }                            |
+       |                                +-----v-----+
+       |                                |           |
+       +--------------------------------+     D     |
+                   MsgAppendResp        |           |
+                                        +-----------+
+
+```
+
+Note:
+
+- `e[x, y]` stands for all the entries within an index range [x, y] (both inclusive)
+- `Commission(target, [x, y])` stands for a job that the delegate should send `e[x, y]` to the target
+
+#### Choose a Delegate
+
+Since a delegate is responsible to send entries to other group members, it must contain entries as much as possible to complete the commission. The leader follows a bunch of rules to choose a delegate of a group:
+
+1. If all the members are requiring snapshots, choose the delegate randomly.
+2. Among the members who are requiring new entries, choose the node satisfies conditions below :
+
+    1. Must be `recent_active`
+    2. The progress state should be `Replicate` but not `paused`
+    3. The progress has the smallest `match_index`
+
+3. If no delegate is picked, the leader does Log Replication itself. Especially, if a group contains the leader it self, no delegate will be set by default except in some cases such as massively large group, which is able to be controlled by upper layer.
```

Comment:
how do we do by the upper layer for the group which contains the leader?

Reply:
It's unnecessary for the upper layer to acknowledge which group contains the leader because only the leader self can choose delegates. And the leader itself must know which group it belongs to.